src/lsp4clj/json_rpc.clj

```diff
@@ -0,0 +1,115 @@
+(ns lsp4clj.json-rpc
+  "Models LSP JSON-RPC as core.async channels of messages (Clojure hashmaps).
+
+  https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#baseProtocol"
+  (:require
+   [camel-snake-kebab.core :as csk]
+   [camel-snake-kebab.extras :as cske]
+   [cheshire.core :as json]
+   [clojure.core.async :as async]
+   [clojure.java.io :as io]
+   [clojure.string :as string]))
+
+(set! *warn-on-reflection* true)
+
+(defn ^:private read-n-chars [^java.io.InputStream input content-length charset-s]
+  (let [buffer (byte-array content-length)]
```

Comment:
Originally I tried to write all this code with a BufferedReader, but couldn't figure out how to do it right. I was having trouble both with the Reader part and the Buffered part.

Readers work at the character level, but we're down at the byte level. Content-Length is specified in bytes, and in order to correctly parse the message we must read that many bytes, exactly, and interpret them in the given charset. (I've renamed this function `read-n-bytes` to make that clearer.) So as we're reading the content, we can't use a Reader (chars), we have to use an InputStream (bytes).

If you want to work at a higher level (for example, to `(read-line reader)` instead of `(read-header-line input)`. You have to switch back and forth between a Reader wrapper and a lower-level InputStream. But that can cause its own problems. If you use Buffered wrappers you don't have control over how many characters/bytes are read off the underlying InputStream. If a buffered wrapper reads too many characters, they're lost. When you switch back to the InputStream, it won't see those lost characters, meaning the next message will be malformed.

I wouldn't be surprised if there's a Java Way to do this, perhaps with Pushback readers. But that's more advanced than I understand. I decided it'd be easier for me to reason about (though more verbose) if I stayed down at the level of individual bytes. 

Anyway, all that to say, I'd welcome advice about how to make this more readable.

Reply:
Ah, it looks like java 9+ has readNBytes https://docs.oracle.com/javase/9/docs/api/java/io/InputStream.html#readNBytes-byte:A-int-int- I can't remember what we're targeting.

Otherwise, this is fine.

Reply:
Oh too bad... readNBytes would be handy. But, we're still [targeting](https://github.com/clojure-lsp/clojure-lsp/blob/aa46e433e9642c1368336130d1dd99cfc68402a8/.github/workflows/ci.yml#L56) Java 8.